### PR TITLE
Use New Routing V1

### LINKED ISSUE #12: Support  Delegated Routing API from IPIP-337
The idea behind `someguy` is to be a simple app that people can run to have delegated routing endpoint that performs DHT+IPNI lookups in parallel.

I believe our ecosystem still needs this: delegated DHT via a HTTP  endpoint is something that will benefit JS implementations running in browser (cc @achingbrain  @tinytb),  mobile browsers and low-power IoT devices (cc @autonome).

## Proposed next steps

Since Reframe was abandoned, `someguy` should be refactored to provide an endpoint compatible with `GET /routing/v1/providers/{CID}` from [IPIP-337](https://github.com/ipfs/specs/pull/337). This allows users to self-host and use it in their web apps.

## TODO 

TBD if we want to do this work without streaming support, or wait until we have it.

- [ ] (MVP) implement `GET /routing/v1/providers/{CID}`  as-is?
- [ ] update API introduced in [IPIP-337](https://github.com/ipfs/specs/pull/337) to support streaming responses (so `someguy` can start responding with providers as it crawls the DHT, and does not block request for multiple seconds)

### LINKED ISSUE #13: support a DHT-only mode
In thinking about what a joint deployment of indexer & dht might look like, we likely want to run instances of accelerated DHT clients that can fill in DHT records in delegated routing responses.

We probably want those responses to take advantage of the same caching as current indexing responses.

The potential difference in latency means that for uncached responses we'll likely want to make sure we split the response into multiple response packets as results return, and parse incrementally on the client side.

### LINKED ISSUE #14: Support DHT-based Peer routing over HTTP
> part of https://github.com/ipfs-shipyard/someguy/issues/13

- [ ] Create IPIP that extends  API introduced in [IPIP-337](https://github.com/ipfs/specs/pull/337) to support [DHT peer records from libp2p spec ](https://github.com/libp2p/specs/tree/master/kad-dht#peer-routing)
  - [ ] wip: https://github.com/ipfs/specs/pull/417
- [ ] Implement support for it in `someguy` (backed by IPFS DHT)

### LINKED ISSUE #15: Support DHT-based IPNS routing
- [ ] Create IPIP that extends  API introduced in [IPIP-337](https://github.com/ipfs/specs/pull/337) to support returning verifiable  [IPNS Record](https://github.com/ipfs/specs/blob/main/ipns/IPNS.md#ipns-record) blobs
  - Key here is for client to be able to verify record themselves
- [ ] Implement support for it in someguy (backed by IPFS DHT)

### PR DESCRIPTION
**Pull request by hacdias**
- [x] Closes #12
- [x] Closes #13
- [x] Closes #14
- [x] Closes #15 (no PubSub though)

Also added some more options to allow users to add more servers to proxy too (per routing type).

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -2,64 +2,75 @@ package main
...
-		dht:     d,
+		cr: crRouters,
+		pr: prRouters,
+		vs: ipnsRouters,
 	}
 
-	f := drs.DelegatedRoutingAsyncHandler(proxy)
-	http.Handle("/", f)
+	log.Printf("Listening on http://0.0.0.0:%d", port)
+	log.Printf("Delegated Routing API on http://127.0.0.1:%d/routing/v1", port)
```
Comment (aschmahmann) [inline server.go:71]: Might be fine, but also a little confusing since routing/v1 is on 127.0.0.1 and all other interfaces too.

**Review by aschmahmann — CHANGES_REQUESTED** [review]
Looks good, thanks 🙏. Haven't gotten a chance to run it and push on it though. Had some small comments and suggestions.

File: main.go [hunk 2]
```diff
@@ -29,21 +29,36 @@ func main() {
 					},
 					&cli.BoolFlag{
 						Name:  "accelerated-dht",
-						Usage: "run the accelerated dht client",
+						Usage: "run the accelerated DHT client",
 						Value: true,
 					},
+					&cli.StringSliceFlag{
+						Name:  "content-endpoints",
```
Comment (lidel) [inline main.go:36]: nit: match endpoint name from /routing/v1 like we do for other two
```suggestion
						Name:  "provider-endpoints",
```

File: go.mod [hunk 3]
```diff
@@ -1,128 +1,122 @@
...
-	github.com/ipfs/go-ipns v0.1.2
-	github.com/ipld/go-ipld-prime v0.17.0
-	github.com/libp2p/go-libp2p v0.21.0
-	github.com/libp2p/go-libp2p-core v0.19.1
-	github.com/libp2p/go-libp2p-kad-dht v0.17.0
-	github.com/libp2p/go-libp2p-record v0.1.3
-	github.com/libp2p/go-libp2p-resource-manager v0.5.1
-	github.com/libp2p/go-libp2p-routing-helpers v0.2.3
-	github.com/urfave/cli/v2 v2.8.1
+	github.com/ipfs/boxo v0.13.2-0.20231009073559-45c797e0ccea
```
[the remark this answers is not in the document]
Comment (hacdias) [reply go.mod:6]: Yeah, the reason I'm not using latest is because `client.Client` is not exposed 🥲

**Review by aschmahmann — APPROVED** [review]
LGTM

**Comment by hacdias** [comment]
Merging for now. Following up on #18 once Boxo is released.

**Pull request merged**